src/queue/multilevel.rs

```diff
@@ -0,0 +1,859 @@
+    pub fn new(config: Config) -> Builder {
+        let manager = Arc::new(LevelManager {
+            level0_elapsed: Default::default(),
+            total_elapsed: Default::default(),
+            task_elapsed_map: Default::default(),
+            level_time_threshold: config.level_time_threshold,
+            level0_chance: AtomicU32::new(INIT_LEVEL0_CHANCE),
+            level0_proportion_target: config.level0_proportion_target,
+            adjusting: AtomicBool::new(false),
+        });
+        Builder { manager }
+    }
+
+    /// Creates a runner builder for the multilevel task queue with a normal runner builder.
+    pub fn runner_builder<B>(&self, inner_runner_builder: B) -> MultilevelRunnerBuilder<B> {
+        MultilevelRunnerBuilder {
+            inner: inner_runner_builder,
+            manager: self.manager.clone(),
+        }
+    }
+
+    fn build_raw<T>(self, local_num: usize) -> (QueueInjector<T>, Vec<QueueLocal<T>>) {
+        let level_injectors: Arc<[Injector<T>; LEVEL_NUM]> =
+            Arc::new(InitWith::init_with(Injector::new));
+        let workers: Vec<_> = iter::repeat_with(Worker::new_lifo)
+            .take(local_num)
+            .collect();
+        let stealers: Vec<_> = workers.iter().map(Worker::stealer).collect();
+        let locals = workers
+            .into_iter()
+            .enumerate()
+            .map(|(self_index, local_queue)| {
+                let stealers = stealers
+                    .iter()
+                    .enumerate()
+                    .filter(|(index, _)| *index != self_index)
+                    .map(|(_, stealer)| stealer.clone())
+                    .collect();
+                QueueLocal {
+                    local_queue,
+                    level_injectors: level_injectors.clone(),
+                    stealers,
+                    manager: self.manager.clone(),
+                    rng: SmallRng::from_rng(thread_rng()).unwrap(),
+                }
+            })
+            .collect();
+
+        (
+            QueueInjector {
+                level_injectors,
+                manager: self.manager,
+            },
+            locals,
+        )
+    }
+
+    /// Creates the injector and local queues of the multilevel task queue.
+    pub fn build<T>(self, local_num: usize) -> (TaskInjector<T>, Vec<LocalQueue<T>>) {
+        let (injector, locals) = self.build_raw(local_num);
+        let local_queues = locals
+            .into_iter()
+            .map(|local| LocalQueue(LocalQueueInner::Multilevel(local)))
+            .collect();
+        (
+            TaskInjector(InjectorInner::Multilevel(injector)),
+            local_queues,
+        )
+    }
+}
+
+thread_local!(static RECENT_NOW: Cell<Instant> = Cell::new(Instant::now()));
+
+/// Returns an instant corresponding to now and updates the thread-local recent
+/// now.
+///
+/// You should only use it when the thread-local recent now is recently updated.
```

Comment:
Shouldn't it be only used when recent now is *not* recently updated?

Reply:
Oh, it's a mistake. It should belong to the `recent` function.